File: network/src/main/kotlin/org/bitlap/network/core/NetworkService.kt
package org.bitlap.network.core

import org.bitlap.network.core.operation.OperationHandle

/**
 * Interface definition of driver RPC.
 * @author 梦境迷离
 * @since 2021/6/6
 * @version 1.0
 */
interface NetworkService {

    fun openSession(username: String, password: String, configuration: Map<String, String>?): SessionHandle

    fun closeSession(sessionHandle: SessionHandle)

    fun executeStatement(
        sessionHandle: SessionHandle,
        statement: String,
        confOverlay: Map<String, String>?
    ): OperationHandle

    fun executeStatement(
        sessionHandle: SessionHandle,
        statement: String,
        confOverlay: Map<String, String>?,
        queryTimeout: Long
    ): OperationHandle

    fun fetchResults(opHandle: OperationHandle): RowSet

    fun getResultSetMetadata(
        opHandle: OperationHandle,
    ): TableSchema

    fun getColumns(
        sessionHandle: SessionHandle,
        tableName: String?,
        schemaName: String?,
        columnName: String?
    ): OperationHandle

    fun getTables(
        sessionHandle: SessionHandle,
        tableName: String?,
        schemaName: String?
    ): OperationHandle

    fun getSchemas(
        sessionHandle: SessionHandle,
    ): OperationHandle
}

File: network/src/main/scala/org/bitlap/network/NetworkService.scala
/* Copyright (c) 2022 bitlap.org */
package org.bitlap.network

import org.bitlap.network.handles.{ OperationHandle, SessionHandle }
import org.bitlap.network.models.{ RowSet, TableSchema }

/**
 * @author 梦境迷离
 * @since 2021/11/20
 * @version 1.0
 */
trait NetworkService {

  def openSession(username: String, password: String, configuration: Map[String, String] = Map.empty): SessionHandle

  def closeSession(sessionHandle: SessionHandle)

  def executeStatement(
    sessionHandle: SessionHandle,
    statement: String,
    confOverlay: Map[String, String]
  ): OperationHandle

  def executeStatement(
    sessionHandle: SessionHandle,
    statement: String,
    queryTimeout: Long,
    confOverlay: Map[String, String] = Map.empty
  ): OperationHandle

  def fetchResults(opHandle: OperationHandle): RowSet

  def getResultSetMetadata(opHandle: OperationHandle): TableSchema

  def getColumns(
    sessionHandle: SessionHandle,
    tableName: String = null,
    schemaName: String = null,
    columnName: String = null
  ): OperationHandle

  def getTables(sessionHandle: SessionHandle, tableName: String = null, schemaName: String = null): OperationHandle

  def getSchemas(sessionHandle: SessionHandle): OperationHandle
}
